<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

# доп. библиотеки
import plotly
import plotly.express as px

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса

# подсчет количества вакансий 
# поле "name" (название вакансии)
# в таблице "vacancies" (данные по вакансиям)
query_3_1 = f'''
    SELECT count(name) as name_count
    FROM public.vacancies
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
print(f'Количество вакансий: {df['name_count'].iloc[0]}')

Количество вакансий: 49197


C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3051625853.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса

# подсчет количества работодателей 
# поле "id" (ID вакансии)
# в таблице "employers" (список работодателей)
query_3_2 = f'''
    SELECT count(id) as id_count
    FROM public.employers
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
print(f'Количество работодателей: {df['id_count'].iloc[0]}')

Количество работодателей: 23501


C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\1166732702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса

# подсчет количества городов 
# поле "id" (идентификатор города)
# в таблице "areas" (список регионов)
query_3_3 = f'''
    SELECT count(id) as id_count
    FROM public.areas
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
print(f'Количество регионов: {df['id_count'].iloc[0]}')

Количество регионов: 1362


C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3269075998.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса

# подсчет количества сфер деятельности 
# поле "id" (ID сферы деятельности)
# в таблице "industries" (список сфер деятельности работодателей)
query_3_4 = f'''
    SELECT count(id) as id_count
    FROM public.industries
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
print(f'Количество сфер деятельности: {df['id_count'].iloc[0]}')

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\24122999.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


Количество сфер деятельности: 294


***

### выводы по предварительному анализу данных
1. Количество анализируемых вакансий 49197.
2. Количество работодателей (23501) вдвое меньше количества вакансий.
3. В анализируемой таблице, присутствуют данные о 294 сферах деятельности в 1362 регионах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса

# подсчет количества вакансий
# соединение таблицы "vacancies" (список вакансий) и таблицы "areas" (список регионов)
# левое соединение по идентификатору города
query_4_1 = f'''
    SELECT a.name as area, count(v.area_id) as cnt 
    FROM public.areas as a 
    LEFT JOIN public.vacancies as v ON a.id = v.area_id
    GROUP BY a.name, a.id
    ORDER BY cnt DESC
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


,area,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892
...,...,...
1357,Балакирево,0
1358,Мексика,0
1359,Пугачев,0
1360,Долгоруково,0


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса

# подсчет количества вакансий
# анализ таблицы "vacancies" (список вакансий)
# условия: поле "Нижняя граница зарплатной вилки" не пусто или поле "Верхняя граница зарплатной вилки" не пусто
query_4_2 = f'''
    SELECT count(id) as id_count
    FROM public.vacancies
    WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
print(f'Количество вакансий: {df['id_count'].iloc[0]}')

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\1858067463.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество вакансий: 24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса

# расчет среднего значения для нижней и верхней зарплатной вилки
# анализ таблицы "vacancies" (список вакансий)
# условия: поле "Нижняя граница зарплатной вилки" не пусто или поле "Верхняя граница зарплатной вилки" не пусто
query_4_3 = f'''
    SELECT 
        round(avg(salary_from)) as salary_from_avg, 
        round(avg(salary_to)) as salary_to_avg
    FROM public.vacancies
    WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
print(f'Среднее значение нижней границы: {df['salary_from_avg'].iloc[0]} руб.')
print(f'Среднее значение верхней границы: {df['salary_to_avg'].iloc[0]} руб.')

Среднее значение нижней границы: 71065.0 руб.
Среднее значение верхней границы: 110537.0 руб.


C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3163178074.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса

# расчет количества вакансий, в зависимости от типа рабочего графика и типа трудоустройства
# анализ таблицы "vacancies" (список вакансий)
# группировка данных по полю "тип рабочего графика" и полю "тип трудоустройства"
query_4_4 = f'''
    SELECT count(id), schedule, employment
    FROM public.vacancies
    GROUP BY schedule, employment
    ORDER BY count(id) DESC
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


,count,schedule,employment
0,35367,Полный день,Полная занятость
1,7802,Удаленная работа,Полная занятость
2,1593,Гибкий график,Полная занятость
3,1312,Удаленная работа,Частичная занятость
4,940,Сменный график,Полная занятость
5,569,Полный день,Стажировка
6,367,Вахтовый метод,Полная занятость
7,347,Полный день,Частичная занятость
8,312,Гибкий график,Частичная занятость
9,141,Полный день,Проектная работа


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса

# расчет количества вакансий, в зависимости от требуемого опыта
# анализ таблицы "vacancies" (список вакансий)
# группировка данных по полю "требования к опыту"
query_4_5 = f'''
    SELECT count(id), experience
    FROM public.vacancies
    GROUP BY experience
    ORDER BY count(id)
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


,count,experience
0,1337,Более 6 лет
1,7197,Нет опыта
2,14511,От 3 до 6 лет
3,26152,От 1 года до 3 лет


***

### выводы по детальному анализу вакансий
1. Наибольшее количество вакансий в Москве и крупных городах. Присутствуют регионы без вакансий, в основном это мелкие регионы.
2. Более половины данных по вакансиям имеют не заполненные границы заработных плат.
3. Наблюдается сильное отличие среднего значения нижней и верхней границы заработных плат. Возможно, данное различие обусловлено разницей в регионах (крупные и мелкие).
4. Наибольшей популярностью пользуются вакансии на полный рабочий день и полную занятость.
5. Наибольший спрос на специалистов с опытом работы от 1 до 3 лет, наименьший - на обладателей большого опыта (более 6 лет).

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса

# расчет количества вакансий, в зависимости от требуемого опыта
# анализ таблицы "employers" (список работодателей) и таблицы "vacancies" (список вакансий)
# левое соединение по идентификатру работодателя
# группировка данных по полю "номер работодателя" и по полю "название работодателя"
query_5_1 = f'''
    SELECT e.id, e.name, count(v.id)
    FROM public.employers as e
    LEFT JOIN public.vacancies as v ON e.id = v.employer_id
    GROUP BY e.id, e.name
    ORDER BY count(v.id) DESC
    LIMIT 5
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


,id,name,count
0,1740,Яндекс,1933
1,2748,Ростелеком,491
2,78638,Тинькофф,444
3,3529,СБЕР,428
4,39305,Газпром нефть,331


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса

# расчет количества вакансий и работодателей, в зависимости от региона
# анализ таблицы "employers" (список работодателей), таблицы "vacancies" (список вакансий) и таблицы "areas" (список регионов)
# левое соединение по идентификатру города (таблица "areas" и таблица "vacancies")
# левое соединение по идентификатру работодателя (таблица "employers" и таблица "vacancies")
# группировка данных по полю "идентификатору города" и по полю "название города"
# поиск регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей
query_5_2 = f'''
    SELECT 
        a.name, 
        count(distinct v.id) as vacancies_count, 
        count(distinct e.id) as employers_count
    FROM public.areas a
    JOIN public.vacancies v ON v.area_id = a.id
    JOIN public.employers e ON v.employer_id = e.id
    GROUP BY a.id, a.name

union all
(
    SELECT 
        a.name, 
        0 as vacancies_count, 
        count(distinct e.id) as employers_count
    FROM public.areas a
    JOIN public.employers e ON e.area = a.id
    WHERE NOT EXISTS
    (
        SELECT 1
            FROM public.vacancies v 
            WHERE v.area_id = a.id
    )
    GROUP BY a.id, a.name
    ORDER BY employers_count DESC
    LIMIT 1
)
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


,name,vacancies_count,employers_count
0,Москва,5333,2320
1,Санкт-Петербург,2851,1451
2,Екатеринбург,1698,742
3,Новосибирск,2006,802
4,Австрия,4,1
...,...,...,...
765,Актогай,1,1
766,Большие Вязёмы,1,1
767,Софьино (Раменский район),2,2
768,Сириус,1,1


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса

# расчет количества регионов, в которых работодатель публикует свои вакансии
# анализ таблицы "employers" (список работодателей), таблицы "vacancies" (список вакансий) и таблицы "areas" (список регионов)
# левое соединение по идентификатру города (таблица "areas" и таблица "vacancies")
# левое соединение по идентификатру работодателя (таблица "employers" и таблица "vacancies")
# группировка данных по полю "номеру работодателя" и по полю "названию работодателя"
query_5_3 = f'''
    SELECT e.name, 
           count(distinct a.id) as areas_count
    FROM public.employers e
    JOIN public.vacancies v ON v.employer_id = e.id
    JOIN public.areas a ON v.area_id = a.id
    GROUP BY e.id, e.name
    ORDER BY areas_count DESC
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


,name,areas_count
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71
...,...,...
14901,НПП Авиатрон,1
14902,Центр дистанционных торгов,1
14903,Городские Телекоммуникационные Системы,1
14904,"Введенский, Отель",1


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса

# расчет количества работодателей, у которых нет сфер деятельности
# анализ таблицы "employers" (список работодателей), таблицы "employers_industries" (связь между работодателями и их сферами деятельности) и таблицы "industries" (справочник вариантов сфер деятельности работодателей)
# левое соединение по номеру работодателя (таблица "employers" и таблица "employers_industries")
# левое соединение по номеру сферы деятельности (таблица "employers" и таблица "industries")
# условие отбора данных: идентификатор сферы деятельности не пусто
query_5_4 = f'''
    SELECT count(*) as "количество работодателей"
    FROM public.employers e
    LEFT JOIN public.employers_industries ei ON ei.employer_id = e.id
    LEFT JOIN public.industries i ON ei.industry_id = i.id
    WHERE i.id IS NULL
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


,количество работодателей
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса

# поиск названия работодателя, у которых четыре сферы деятельности
# анализ таблицы "employers" (список работодателей), таблицы "employers_industries" (связь между работодателями и их сферами деятельности) и таблицы "industries" (справочник вариантов сфер деятельности работодателей)
# левое соединение по номеру работодателя (таблица "employers" и таблица "employers_industries")
# левое соединение по номеру сферы деятельности (таблица "employers" и таблица "industries")
# группировка данных по номеру и названию работодателя
# оставляем только группы, в которых количество записей 4
query_5_5 = f'''
    SELECT e.name
    FROM public.employers e
    JOIN public.employers_industries ei ON ei.employer_id = e.id
    JOIN public.industries i ON ei.industry_id = i.id
    GROUP BY e.name, e.id
    HAVING count(*) = 4
    ORDER BY lower(e.name)
    LIMIT 3
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


,name
0,101 Интернет
1,21vek.by
2,2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса

# поиск количества работодателя, у которых сфера деятельности "Разработка программного обеспечения"
# анализ таблицы "employers" (список работодателей), таблицы "employers_industries" (связь между работодателями и их сферами деятельности) и таблицы "industries" (справочник вариантов сфер деятельности работодателей)
# левое соединение по номеру работодателя (таблица "employers" и таблица "employers_industries")
# левое соединение по номеру сферы деятельности (таблица "employers" и таблица "industries")
query_5_6 = f'''
    SELECT count(distinct e.id) as employers_count
    FROM public.employers e
    JOIN public.employers_industries ei ON ei.employer_id = e.id
    JOIN public.industries i ON ei.industry_id = i.id AND i.name ILIKE 'Разработка программного обеспечения'
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


,employers_count
0,3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# список городов-милионников
# данные берутся из таблицы сайта https://ru.wikipedia.org/wiki/Города-миллионеры_России

import requests
import mwclient
from bs4 import BeautifulSoup

# URL городов-миллионников
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
# определение user-agent
headers = {"User-Agent": "Mozilla/5.0"}
# GET-запрос к ресурсу и результат ответа сохраняем в переменной response
response = requests.get(url, headers = headers)

# создаём объект BeautifulSoup, указывая html.parser
soup = BeautifulSoup(response.text, 'html.parser')

# Находим таблицу с классом "standard"
table = soup.find('table', {'class': 'standard'})

#сбор данных городов-миллионников
towns_million = []
if table:
    for row in table.find_all('tr'):
        cells = row.find_all('td')
        if len(cells) >= 2:
            city_link = cells[1].find('a')
            if city_link:
                towns_million.append(city_link.get_text(strip = True))
else:
    print('Таблица с городами не найдена')

# получение списка городов-миллионников
towns_million_sql = ", ".join([f"'{i}'" for i in towns_million])
print(towns_million_sql)

'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'


In [35]:
# текст запроса

# подсчет количества вакансий для компании "Яндекс" в городах-миллионниках
# анализ таблицы "employers" (список работодателей), таблицы "vacancies" (список вакансий) и таблицы "areas" (список регионов)
# левое соединение по номеру работодателя (таблица "employers" и таблица "vacancies")
# левое соединение по идентификатору города (таблица "employers" и таблица "areas")
# условие отбора данных: поле "название работодателя" содержит "Яндекс" и поле "название города" содержится в списке городов-миллионников
query_5_7 = f'''
    WITH v_vacancies_million as
    (
        SELECT 
            a.name as area_name, 
            count(*) as count
        FROM public.employers e
        JOIN public.vacancies v ON v.employer_id = e.id
        JOIN public.areas a ON a.id = v.area_id
        WHERE 
            e.name ILIKE 'Яндекс' AND a.name IN ({towns_million_sql})
        GROUP BY e.id, a.id, a.name
    )

    (SELECT 
        vm.area_name as "город миллионник", 
        vm.count as "кол-во вакансий"
    FROM v_vacancies_million vm 
    ORDER BY count DESC)
    UNION ALL
    (SELECT 
        'Total' as area_name, 
        sum(vm.count) as count 
    FROM v_vacancies_million vm 
    ORDER BY count DESC)
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


,город миллионник,кол-во вакансий
0,Москва,54.0
1,Санкт-Петербург,42.0
2,Екатеринбург,39.0
3,Нижний Новгород,36.0
4,Новосибирск,35.0
5,Воронеж,32.0
6,Краснодар,30.0
7,Самара,26.0
8,Уфа,26.0
9,Казань,25.0


***

### выводы по анализу работодателей
1. Больше всего вакансий к компании Яндекс.
2. В Москве, Санкт-Петербурге больше всего работодателей и вакансий. Реже всего вакансий в регионах.
3. Компании Яндекс, Ростелеком, Спецремонт больше всего присутствуют в регионах.
4. У 8419 работодателей не указана сфера деятельности.
5. Компания 2ГИС находится на третьем месте среди компаний у которых указаны 4 сферы деятельности.
6. 3553 компаний указали разработку программного обеспечения в сфере своей деятельности.
7. Компания Яндекс предствалена в городах миллионниках, общее количество вакансий 485.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса

# подсчет количества вакансий, в наименовании которых содержится слово "data" или "данн"
# анализ таблицы "vacancies" (список вакансий)
query_6_1 = f'''
    SELECT count(*) as vacancies_count
    FROM public.vacancies v
    WHERE v.name ilike '%data%' OR v.name ilike '%данн%'
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


,vacancies_count
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса

# подсчет количества вакансий, в наименовании которых содержится слово:
# анализ таблицы "vacancies" (список вакансий)
# * 'data scientist'
# * 'data science'
# * 'исследователь данных'
# * 'ML' (здесь не нужно брать вакансии по HTML)
# * 'machine learning'
# * 'машинн%обучен%'
# анализ таблицы "vacancies" (список вакансий)
query_6_2 = f'''
    SELECT count(v.name) as vacancies_count
    FROM public.vacancies v
    WHERE
        (
            v.name ILIKE '%data scientist%'
            OR v.name ILIKE '%data science%'
            OR v.name ILIKE '%исследователь данных%'
            OR ((v.name ILIKE '%ML%') and (name NOT ILIKE '%HTML%'))
            OR v.name ILIKE '%machine learning%'
            OR v.name ILIKE '%машинн%обучен%'
        )
        AND
        (
            v.name ILIKE '%junior%'
            OR v.experience ILIKE '%Нет опыта%'
            OR v.employment ILIKE '%Стажировка%'
        )
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


,vacancies_count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса

# подсчет количества вакансий 
# анализ таблицы "vacancies" (список вакансий)
# в наименовании которых содержится слово: 'data scientist', 'data science', 'исследователь данных', 'machine learning', 'машинн%обучен', 'ML', 'HTML'
# в наименовании ключевых навыков содержится слово "SQL" или "postgres"
query_6_3 = f'''
    SELECT count(v.name) as vacancies_count
    FROM public.vacancies v
    WHERE
        (
            1=0
            OR name ILIKE '%data scientist%'
            OR name ILIKE '%data science%'
            OR name ILIKE '%исследователь данных%'
            OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
            OR name ILIKE '%machine learning%'
            OR name ILIKE '%машинн%обучен%'
        )
        and
        (
            v.key_skills ILIKE '%SQL%'
            OR v.key_skills ILIKE '%postgres%'
        )
'''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


,vacancies_count
0,229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса

# подсчет количества вакансий 
# анализ таблицы "vacancies" (список вакансий)
# в наименовании которых содержится слово: 'data scientist', 'data science', 'исследователь данных', 'machine learning', 'машинн%обучен', 'ML', 'HTML'
# в наименовании ключевых навыков содержится слово "SQL" или "postgres"
query_6_4 = f'''
    SELECT count(v.name) as vacancies_count
    FROM public.vacancies v
    WHERE
        (
            name ILIKE '%data scientist%'
            OR name ILIKE '%data science%'
            OR name ILIKE '%исследователь данных%'
            OR ((name ILIKE '%ML%') and (name NOT ILIKE '%HTML%'))
            OR name ILIKE '%machine learning%'
            OR name ILIKE '%машинн%обучен%'
        )
        AND
        (
            v.key_skills ILIKE '%python%'
        )
'''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


,vacancies_count
0,357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса

# подсчет среднего количества ключевых навыков
# анализ таблицы "vacancies" (список вакансий)
# в наименовании которых содержится слово: 'data scientist', 'data science', 'исследователь данных', 'machine learning', 'машинн%обучен', 'ML'
# в наименовании которых не содержится слово: 'HTML'

query_6_5 = f'''
    SELECT round(avg(length(v.key_skills) - length(replace(key_skills, chr(9), '')) + 1), 2) as vacancies_count
    FROM public.vacancies v
    WHERE
        (
            name ILIKE '%data scientist%'
            OR name ILIKE '%data science%'
            OR name ILIKE '%исследователь данных%'
            OR name LIKE '%ML%'
            OR name ILIKE '%machine learning%'
            OR name ILIKE '%машинн%обучен%'
            OR name LIKE '%ML%'
        )
        AND name NOT ILIKE '%HTML%'
'''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


,vacancies_count
0,6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса

# подсчет средней зарплаты для каждого типа требуемого опыта
# анализ таблицы "vacancies" (список вакансий)
# в наименовании которых содержится слово: 'data scientist', 'data science', 'исследователь данных', 'machine learning', 'машинн%обучен', 'ML'
# в наименовании которых не содержится слово: 'HTML'
# требуемый опыт содержит значение от 3 до 6 лет
query_6_6 = f'''
    SELECT 
        round
        (
            avg
            ( 
                case
                    WHEN v.salary_from IS NOT NULL AND v.salary_to IS NOT NULL THEN (v.salary_from + v.salary_to) / 2
                    WHEN v.salary_from IS NOT NULL THEN v.salary_from
                    ELSE v.salary_to
                end
            )
            , 0
        )::int as salary_avg
    FROM public.vacancies v
    WHERE
        (
            1=0
            OR name ILIKE '%data scientist%'
            OR name ILIKE '%data science%'
            OR name ILIKE '%исследователь данных%'
            OR name LIKE '%ML%'
            OR name ILIKE '%machine learning%'
            OR name ILIKE '%машинн%обучен%'
        )
        AND name NOT ILIKE '%HTML%'
        AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
        AND v.experience = 'От 3 до 6 лет'
'''

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


,salary_avg
0,243115


***

### выводы по предметному анализу
1. Проанализировано 49197 вакансий.
2. 1771 вакансий (около 3.6%) связаны с "data" или "данными".
3. Наблюдается низкое количество вакансий для junior-специалистов (51).
4. Популярные навыки: SQL (229 вакансий), Python (357 вакансий).
5. В среднем в вакансиях указывается 6 ключевых навыков.
6. Средняя заработная плата составляет 243115 рублей.
7. Рекомендуется развивать навыки в области Data, в частности, SQL и Python.

# Общий вывод по проекту

### подведем итог исследования, обобщите выводы

1. В больших городах вакансий больше. Как правило, это Москва и Санкт-Петербург.
2. Наибольший запрос вакансий предоставляют крупные работодатели федерального уровня.
3. Чем больше опыт, тем ниже текучка кадров, максимальное количество вакансий для соискателей с небольшим опытом. 
4. Больше всего вакансий на полную занятость и полный день. 
5. Средняя заработная плата для специалистов data scientis с опытом от 3 до 6 лет выше, чем в среднем по IT.
6. 3.6% вакансий напрямую связаны с "data" или "данными"

### здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Ниже представлены графики для наглядного анализа  
Исследование можно продолжить, углубив анализ в следующих направлениях:  
1. анализ самых популярных вакансий
2. анализ самых популярных вакансий и работодателей по регионам
3. средняя заработная плата по регионам
4. средняя заработная плата по вакансиям

In [49]:
# Дополнительное исследование 1 (анализ самых популярных вакансий)

# выборка 10 самых популярных вакансий
# анализ таблицы "vacancies" (список вакансий)
query = f'''
    SELECT 
        replace(name, '1C', '1С') as name,
        count(*) as count
    FROM public.vacancies
    WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
    GROUP BY name
    ORDER BY count DESC
    LIMIT 10
'''

# запрос
df = pd.read_sql_query(query, connection)

# график: столбчатая диграмма
fig = px.bar(
    data_frame = df, # источник данных
    x = 'count', # ось абсцисс (x), количество вакансий
    y = 'name',  # ось ординат (y), название вакансий
    title = 'Топ 10 вакансий', # заголовок графика
    color = 'count', # раскраска
    labels = {'count': 'количество'}, # легенда
    text_auto = True # подписи величин
)

fig.update_layout(
    xaxis_title = 'Количество вакансий', # ось абсцисс (x)
    yaxis_title = 'Название вакансии', # ось ординат (y)
    title_x = 0.5,
    showlegend = False
)

fig.show(width = 1000, height = 800)

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3493341740.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


Выводы:  
на графике видно, что лидирующие позиции, в части ваканcий на рынке труда, занимимает сфера IT

In [50]:
# Дополнительное исследование 2 (анализ самых популярных вакансий и работодателей по регионам)

# выборка 10 самых популярных регионов, по количеству вакансий
# анализ таблицы "vacancies" (список вакансий) и таблицы "areas" (список регионов)
# левое соединение по идентификатору города
query_areas_top_vac = f'''
    SELECT a.name, count(*) AS vacancy_count
    FROM areas a
    JOIN vacancies v ON a.id = v.area_id
    GROUP BY a.name
    ORDER BY vacancy_count DESC
    LIMIT 10
'''

# запрос (10 самых популярных регионов, по количеству вакансий)
df_areas_top_vac = pd.read_sql_query(query_areas_top_vac, connection)

# выборка 10 самых популярных работодателей, по регионам
query_areas_top_employers = f'''
    SELECT a.name, count(DISTINCT e.id) AS employer_count
    FROM areas a
    JOIN vacancies v ON a.id = v.area_id
    JOIN employers e ON v.employer_id = e.id
    GROUP BY a.name
    ORDER BY employer_count DESC
    LIMIT 10
'''

# запрос (10 самых популярных работодателей, по регионам)
df_areas_top_employers = pd.read_sql_query(query_areas_top_employers, connection)

# соединяем данные по выборкам
data = pd.merge(df_areas_top_vac, 
                df_areas_top_employers, 
                on = 'name', 
                how = 'outer').fillna(0).sort_values('vacancy_count', ascending = False)

# переименование столбцов
data = data.rename(columns = {'name': 'Region',
                               'vacancy_count' : 'Количество вакансий',
                               'employer_count' : 'Количество работодателей'})
# сводная таблица
data_melt = pd.melt(data, 
                    id_vars = ['Region'],
                    value_vars = ['Количество вакансий', 'Количество работодателей'],
                    var_name = 'легенда', 
                    value_name = 'Count')

# график: столбчатая диграмма
fig = px.bar( 
    data_frame = data, 
    x = ['Количество вакансий','Количество работодателей'], # ось абсцисс (x), количество вакансий/работодателей
    y = data['Region'],  # ось ординат (y), регионы
    #barmode = 'Region', # расположение групп, по горизонтали
    title = 'Количество вакансий и работодателей по регионам',
    labels = {'variable': 'обозначение: '}, # легенда
    width = 500,  # ширина графика
    height = 400, # высота графика
    text_auto = True,
)  

# настройки графика
fig.update_layout(
    xaxis_title = 'регионы', # ось абсцисс (x)
    yaxis_title = 'количество вакансий/работодателей', # ось ординат (y)
    legend = dict(orientation = 'h', # расположение легенды
                  x = 0, 
                  y = 150, 
                  title_font_family = 'Sitka Small') 
)

fig.show()

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3402711486.py:16: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3402711486.py:30: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Выводы:  
из графика видно, что подавляющее число вакансий и работодателей находится в Москве. Далее следует Санкт-Петербург и Минск.

In [51]:
# Дополнительное исследование 3 (средняя заработная плата по регионам)

# выборка 10 самых популярных регионов, по размеру средней заработной платы
# анализ таблицы "vacancies" (список вакансий) и таблицы "areas" (список регионов)
# левое соединение по идентификатору города
query_area = f'''
    SELECT a.name AS region,
    ROUND(AVG(COALESCE((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0) AS avg_salary
    FROM vacancies v
    JOIN areas a ON v.area_id = a.id
    WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
    GROUP BY a.name
    ORDER BY avg_salary DESC
    LIMIT 10
'''

# запрос (10 самых популярных регионов, по размеру средней заработной платы)
df_areas_top_zp = pd.read_sql_query(query_area, connection)

# график: столбчатая диграмма
fig = px.bar(
    data_frame = df_areas_top_zp,
    #x = 'region',   # ось абсцисс (x), регионы
    #y = 'avg_salary', # ось ординат (y), средняя заработная плата (руб.)
    x = 'avg_salary',   # ось абсцисс (x), регионы
    y = 'region', # ось ординат (y), средняя заработная плата (руб.)
    text_auto = True, # отображение надписей, на колонках гистограммы
    title = 'Средняя заработная плата по регионам, руб.', # заголовок
    width = 1000,    # ширина графика
    height = 500,    # высота графика
    text = 'region',  # вывод регионы
    color = 'region', # расраска по регионам
)

# отображение графика (визуализация)
fig.update_layout(
    bargap = 0.1, # ширина вертикальной разделительной полосы
    xaxis_title = 'Регионы', # подпись оси абсции (x)
    yaxis_title = 'Средняя заработная плата, руб.', # подпись оси ординат (y)
)

fig.show()

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3400991021.py:18: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Выводы:  
из графика видно, что за рубежом средняя заработная плата выше

In [52]:
# Дополнительное исследование 4 (средняя заработная плата по вакансиям)

# выборка 10 самых популярных вакансий, по размеру средней заработной платы
# анализ таблицы "vacancies" (список вакансий)
query_avg_salary = f'''
    SELECT name AS vacancies, ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) AS avg_salary
    FROM vacancies v
    WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL
    GROUP BY 1
    ORDER BY avg_salary DESC
    LIMIT 10
'''

# запрос
df = pd.read_sql_query(query_avg_salary, connection)

# график: столбчатая диграмма
fig = px.bar(
    data_frame = df, # источник данных
    x = 'avg_salary', # ось абсцисс (x), средняя заработная плата
    y = 'vacancies',  # ось ординат (y), название вакансий
    title = 'Средняя заработная плата по вакансиям, руб.', # заголовок графика
    #color = 'count', # раскраска
    #labels = {'count': 'количество'}, # легенда
    text_auto = True # подписи величин
)

fig.update_layout(
    xaxis_title = 'Средняя заработная плата', # ось абсцисс (x)
    yaxis_title = 'Название вакансии', # ось ординат (y)
    title_x = 0.5,
    showlegend = False
)

fig.show()

C:\Users\Ирина\AppData\Local\Temp\ipykernel_6956\3165969717.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Выводы:  
из графика видно, что самая большая заработная плата у работников IT/разработчиков.

In [53]:
# закрываем соединение
connection.close()